include/revng/Model/NameBuilder.h

```diff
@@ -199,6 +202,189 @@ class NameBuilder {
     return Result;
   }
 
+  /// This helper provides shared functionality for all the different types
+  /// of counting name builders (those that emit `prefix_1`, `prefix_2`,
+  /// `prefix_3`, and so on as default names on sequential requests).
+  ///
+  /// It ensures that:
+  /// - a single *name* cannot appear more than once
+  /// - a single *location* cannot appear more than once
+  /// - the indexes (either a part of a name, or the corresponding locations)
+  ///   are unique for each \ref nameImpl call.
+  struct CountingNameBuilder {
+  protected:
+    const NameBuilder *Parent = nullptr;
+
+  private:
+    uint64_t NextIndex = 0;
+
+    /// This is used to keep track of the *names* that were already emitted.
+    std::unordered_set<std::string_view> EmittedNames = {};
+
+    /// This is used to keep track of the *locations* that were already emitted.
+    std::unordered_set<std::string> EmittedLocations = {};
+
+  public:
+    CountingNameBuilder() = default;
```

Comment:
Why do we expose this?

Reply:
Because these are one-per-function. And backends (both of them, I think, but I'd have to check to be sure) want a single variable they just keep reassigning when needed.

Reply:
Mmmh, OK, maybe leave a `// TODO: consider recreating this for each function`.

Reply:
Done.

BTW, this is for the new backend _only_. But refactoring it doesn't look trivial - there are a lot of recursive coroutines floating around - and it used like _the_ place to reliably put the object.